Cannot use cffi.from_buffer when explicitly passing signature to JIT

I was experimenting with passing arrays to CFFI functions (see https://github.com/numba/numba/pull/1464) but found that it only works when not specifying the signature. This change makes the test fail in 0.23.1:

```diff
diff --git a/numba/tests/test_cffi.py b/numba/tests/test_cffi.py
index ca304f9..fa1752d 100644
--- a/numba/tests/test_cffi.py
+++ b/numba/tests/test_cffi.py
@@ -82,7 +82,9 @@ class TestCFFI(TestCase):
         np.testing.assert_equal(pyfunc(x), cfunc(x))
 
     def test_pass_numpy_array_float32(self):
-        self._test_pass_numpy_array(vector_sin_float32, np.float32)
+        x = np.arange(10).astype(np.float32)
+        cfunc = jit('float32[:](float32[:])', nopython=True)(vector_sin_float32)
+        np.testing.assert_equal(vector_sin_float32(x), cfunc(x))
 
     def test_pass_numpy_array_float64(self):
         self._test_pass_numpy_array(vector_sin_float64, np.float64)
```

I was not able to try it with latest numba since it requires llvmlite 0.9.0 and it's not available in conda yet.

The workaround for me has been to provide a separate function without signature, but it's a pity because it is preventing me to provide different specializations for CFFI functions (see https://github.com/Pybonacci/cffi_test/).

By the way, I forgot to paste the error message:

```bash
$ py.test test_cffi.py 
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.5.1, pytest-2.8.5, py-1.4.31, pluggy-0.3.1
rootdir: /home/juanlu/Development/Python/numba, inifile: 
collected 11 items 

test_cffi.py .F.........

=============================================================================== FAILURES ===============================================================================
________________________________________________________________ TestCFFI.test_pass_numpy_array_float32 ________________________________________________________________

self = <numba.tests.test_cffi.TestCFFI testMethod=test_pass_numpy_array_float32>

    def test_pass_numpy_array_float32(self):
        x = np.arange(10).astype(np.float32)
>       cfunc = jit('float32[:](float32[:])', nopython=True)(vector_sin_float32)

test_cffi.py:86: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../decorators.py:168: in wrapper
    disp.compile(sig)
../dispatcher.py:349: in compile
    flags=flags, locals=self.locals)
../compiler.py:665: in compile_extra
    return pipeline.compile_extra(func)
../compiler.py:366: in compile_extra
    return self.compile_bytecode(bc, func_attr=self.func_attr)
../compiler.py:375: in compile_bytecode
    return self._compile_bytecode()
../compiler.py:652: in _compile_bytecode
    return pm.run(self.status)
../compiler.py:251: in run
    raise patched_exception
../compiler.py:243: in run
    res = stage()
../compiler.py:463: in stage_nopython_frontend
    self.locals)
../compiler.py:780: in type_inference_stage
    infer.propagate()
../typeinfer.py:565: in propagate
    raise errors[0]
../typeinfer.py:111: in propagate
    constraint(typeinfer)
../typeinfer.py:284: in __call__
    self.resolve(typeinfer, typevars, fnty)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <numba.typeinfer.CallConstraint object at 0x7fa6c059e860>, typeinfer = <numba.typeinfer.TypeInferer object at 0x7fa6c0510c50>
typevars = {'$0.1': $0.1 := Module(<module 'numpy' from '/home/juanlu/.miniconda3/envs/numba_test/lib/python3.5/site-packages/num...er for ffi), 'x': x := array(float32, 1d, A), 'y': y := array(float32, 1d, C), '$0.19': $0.19 := array(float32, 1d, A)}
fnty = BoundFunction(ffi.from_buffer for ffi)

    def resolve(self, typeinfer, typevars, fnty):
        assert fnty
        context = typeinfer.context
    
        n_pos_args = len(self.args)
        kwds = [kw for (kw, var) in self.kws]
        argtypes = [typevars[a.name] for a in self.args]
        argtypes += [typevars[var.name] for (kw, var) in self.kws]
        if self.vararg is not None:
            argtypes.append(typevars[self.vararg.name])
    
        if not all(a.defined for a in argtypes):
            # Cannot resolve call type until all argument types are known
            return
    
        args = tuple(a.getone() for a in argtypes)
        pos_args = args[:n_pos_args]
        if self.vararg is not None:
            if not isinstance(args[-1], types.BaseTuple):
                # Unsuitable for *args
                # (Python is more lenient and accepts all iterables)
                return
            pos_args += args[-1].types
            args = args[:-1]
        kw_args = dict(zip(kwds, args[n_pos_args:]))
        sig = context.resolve_function_type(fnty, pos_args, kw_args)
        if sig is None:
            desc = context.explain_function_type(fnty)
            headtemp = "Invalid usage of {0} with parameters ({1})"
            head = headtemp.format(fnty, ', '.join(map(str, args)))
            msg = '\n'.join([head, desc])
>           raise TypingError(msg, loc=self.loc)
E           numba.errors.TypingError: Failed at nopython (nopython frontend)
E           Invalid usage of BoundFunction(ffi.from_buffer for ffi) with parameters (array(float32, 1d, A))
E            * parameterized
E           File "cffi_usecases.py", line 130

../typeinfer.py:317: TypingError
================================================================= 1 failed, 10 passed in 1.39 seconds ==================================================================
```

--- a/docs/source/reference/pysupported.rst
+++ b/docs/source/reference/pysupported.rst
@@ -374,7 +374,7 @@ Third-party modules
 --------
 
 Similarly to ctypes, Numba is able to call into `cffi`_-declared external
-functions, using the following C types:
+functions, using the following C types and any derived pointer types:
 
 * :c:type:`char`
 * :c:type:`short`
@@ -396,18 +396,15 @@ functions, using the following C types:
 * :c:type:`uint64_t`
 * :c:type:`float`
 * :c:type:`double`
-* :c:type:`char *`
-* :c:type:`void *`
-* :c:type:`uint8_t *`
-* :c:type:`float *`
-* :c:type:`double *`
 * :c:type:`ssize_t`
 * :c:type:`size_t`
 * :c:type:`void`
 
-The ``from_buffer`` method of ``cffi.FFI`` and ``CompiledFFI`` objects is
-supported for passing NumPy arrays of ``float32`` and ``float64`` values to C
-function parameters of type ``float *`` and ``double *`` respectively.
+The ``from_buffer()`` method of ``cffi.FFI`` and ``CompiledFFI`` objects is
+supported for passing Numpy arrays and other buffer-like objects.  Only
+*contiguous* arguments are accepted.  The argument to ``from_buffer()``
+is converted to a raw pointer of the appropriate C type (for example a
+``double *`` for a ``float64`` array).
 
 Out-of-line cffi modules must be registered with Numba prior to the use of any
 of their functions from within Numba-compiled functions:

--- a/numba/targets/cffiimpl.py
+++ b/numba/targets/cffiimpl.py
@@ -10,7 +10,7 @@ from . import arrayobj
 
 registry = Registry()
 
-@registry.lower('ffi.from_buffer', types.Array)
+@registry.lower('ffi.from_buffer', types.Buffer)
 def from_buffer(context, builder, sig, args):
     assert len(sig.args) == 1
     assert len(args) == 1

--- a/numba/tests/cffi_usecases.py
+++ b/numba/tests/cffi_usecases.py
@@ -141,12 +141,8 @@ def use_user_defined_symbols():
 # (cffi_usecases_ool.ffi is a CompiledFFI object) so we use both in these
 # functions.
 
-def vector_sin_float32(x):
-    y = np.empty_like(x)
+def vector_sin_float32(x, y):
     vsSin(len(x), ffi.from_buffer(x), ffi_ool.from_buffer(y))
-    return y
 
-def vector_sin_float64(x):
-    y = np.empty_like(x)
+def vector_sin_float64(x, y):
     vdSin(len(x), ffi.from_buffer(x), ffi_ool.from_buffer(y))
-    return y

--- a/numba/tests/test_cffi.py
+++ b/numba/tests/test_cffi.py
@@ -1,9 +1,11 @@
 from __future__ import print_function, division, absolute_import
 
+import array
 import numpy as np
+import sys
 
 from numba import unittest_support as unittest
-from numba import jit, cffi_support, types
+from numba import jit, cffi_support, types, errors
 from numba.compiler import compile_isolated, Flags
 from numba.tests.support import TestCase
 
@@ -86,16 +88,41 @@ class TestCFFI(TestCase):
         cfunc = jit(nopython=True)(pyfunc)
         self.assertEqual(pyfunc(), cfunc())
 
-    def _test_pass_numpy_array(self, pyfunc, dtype):
+    def check_vector_sin(self, cfunc, x, y):
+        cfunc(x, y)
+        np.testing.assert_allclose(y, np.sin(x))
+
+    def _test_from_buffer_numpy_array(self, pyfunc, dtype):
         x = np.arange(10).astype(dtype)
+        y = np.zeros_like(x)
         cfunc = jit(nopython=True)(pyfunc)
-        np.testing.assert_equal(pyfunc(x), cfunc(x))
+        self.check_vector_sin(cfunc, x, y)
 
-    def test_pass_numpy_array_float32(self):
-        self._test_pass_numpy_array(mod.vector_sin_float32, np.float32)
+    def test_from_buffer_float32(self):
+        self._test_from_buffer_numpy_array(mod.vector_sin_float32, np.float32)
 
-    def test_pass_numpy_array_float64(self):
-        self._test_pass_numpy_array(mod.vector_sin_float64, np.float64)
+    def test_from_buffer_float64(self):
+        self._test_from_buffer_numpy_array(mod.vector_sin_float64, np.float64)
+
+    @unittest.skipIf(sys.version_info < (3,),
+                     "buffer protocol on array.array needs Python 3+")
+    def test_from_buffer_pyarray(self):
+        pyfunc = mod.vector_sin_float32
+        cfunc = jit(nopython=True)(pyfunc)
+        x = array.array("f", range(10))
+        y = array.array("f", [0] * len(x))
+        self.check_vector_sin(cfunc, x, y)
+
+    def test_from_buffer_error(self):
+        pyfunc = mod.vector_sin_float32
+        cfunc = jit(nopython=True)(pyfunc)
+        # Non-contiguous array
+        x = np.arange(10).astype(np.float32)[::2]
+        y = np.zeros_like(x)
+        with self.assertRaises(errors.TypingError) as raises:
+            cfunc(x, y)
+        self.assertIn("from_buffer() unsupported on non-contiguous buffers",
+                      str(raises.exception))
 
 
 if __name__ == '__main__':

--- a/numba/typing/cffi_utils.py
+++ b/numba/typing/cffi_utils.py
@@ -9,6 +9,7 @@ from types import BuiltinFunctionType
 import ctypes
 
 from numba import types
+from numba.errors import TypingError
 from . import templates
 
 try:
@@ -63,7 +64,7 @@ def _type_map():
     global _cached_type_map
     if _cached_type_map is None:
         _cached_type_map = {
-            ffi.typeof('char') :                types.int8,
+            ffi.typeof('char') :                types.char,
             ffi.typeof('short') :               types.short,
             ffi.typeof('int') :                 types.intc,
             ffi.typeof('long') :                types.long_,
@@ -83,11 +84,6 @@ def _type_map():
             ffi.typeof('uint64_t') :            types.ulonglong,
             ffi.typeof('float') :               types.float_,
             ffi.typeof('double') :              types.double,
-            ffi.typeof('char *') :              types.voidptr,
-            ffi.typeof('void *') :              types.voidptr,
-            ffi.typeof('uint8_t *') :           types.CPointer(types.uint8),
-            ffi.typeof('float *') :             types.CPointer(types.float32),
-            ffi.typeof('double *') :            types.CPointer(types.float64),
             ffi.typeof('ssize_t') :             types.intp,
             ffi.typeof('size_t') :              types.uintp,
             ffi.typeof('void') :                types.void,
@@ -101,20 +97,24 @@ def map_type(cffi_type):
     """
     kind = getattr(cffi_type, 'kind', '')
     if kind in ('struct', 'union'):
-        raise TypeError("No support for struct or union")
+        raise TypeError("No support for CFFI %s" % (kind,))
     elif kind == 'function':
         if cffi_type.ellipsis:
             raise TypeError("vararg function is not supported")
         restype = map_type(cffi_type.result)
         argtypes = [map_type(arg) for arg in cffi_type.args]
         return templates.signature(restype, *argtypes)
+    elif kind == 'pointer':
+        pointee = cffi_type.item
+        if pointee.kind == 'void':
+            return types.voidptr
+        else:
+            return types.CPointer(map_type(pointee))
     else:
         result = _type_map().get(cffi_type)
-
-    if result is None:
-        raise TypeError(cffi_type)
-
-    return result
+        if result is None:
+            raise TypeError(cffi_type)
+        return result
 
 
 def make_function_type(cffi_func):
@@ -147,11 +147,15 @@ class FFI_from_buffer(templates.AbstractTemplate):
     key = 'ffi.from_buffer'
 
     def generic(self, args, kws):
-        if kws or (len(args) != 1):
+        if kws or len(args) != 1:
             return
         [ary] = args
-        if not (isinstance(ary, types.Array) and ary.layout in ('C', 'F')):
-            return
+        if not isinstance(ary, types.Buffer):
+            raise TypingError("from_buffer() expected a buffer object, got %s"
+                              % (ary,))
+        if ary.layout not in ('C', 'F'):
+            raise TypingError("from_buffer() unsupported on non-contiguous buffers (got %s)"
+                              % (ary,))
         ptr = types.CPointer(ary.dtype)
         return templates.signature(ptr, ary)
 

